### PR TITLE
fileutils: Avoid compiling a regexp for simple patterns

### PR DESCRIPTION
**Pull request by aaronlehmann**
If we detect that a pattern is either an exact match, prefix match, or
suffix match, use an optimized code path instead of compiling a regexp.

### CODE REVIEW

**Comment by thaJeztah** [comment]
Windows failure is unrelated (I'll kick CI again);

```
=== RUN   TestDockerSuite/TestSlowStdinClosing
    docker_cli_run_test.go:4177: assertion failed: error is not nil: exit status 127
    --- FAIL: TestDockerSuite/TestSlowStdinClosing (4.52s)
```

**Comment by aaronlehmann** [comment]
Review ping?

**Comment by tianon** [comment]
Sorry Aaron - we discussed this a little in the maintainers meeting last week, but forgot to leave some kind of comment.

I think we're a little bit worried about accidentally breaking edge cases, since the test coverage here is pretty decent for individual patterns, but is _not_ very comprehensive for multi-pattern matches (even with this PR doubling the number, we've only got four :weary:).

Speaking less collectively, I wonder if there's maybe a library of existing `.gitignore` tests somewhere (which might have more complex combinations) we could co-opt, obviously ignoring any cases in it that might be Git-specific?  Does Git itself have any useful `.gitignore` tests?  (I think we've tried to make sure `.dockerignore` parses very similarly to `.gitignore`, so finding some existing tests for the latter would help us validate/verify that, especially over time. :sweat_smile:)

**Comment by aaronlehmann** [comment]
BTW, @tonistiigi requested I implement this optimization. I don't personally care much about it, but I did put in the time to do this as a favor, so it would be a bit disappointing if it doesn't move forward.

**Comment by tonistiigi** [comment]
The question wasn't this PR specifically but the previous ones that are already merged and exposed some issues(including the one fixed by current PR). We were discussing how to make sure that the previous ones are good now and what additional testing might be needed to prove that. I'll try to review this soon.

**Review by tonistiigi — COMMENTED** [review]
Could we add a test case that just calls `compile()` on a bunch of inputs and verifies that it was detected as a correct type. It feels that if there is a breakage in there it could easily go unnoticed because another type could still give a correct match result, but inefficiently.

**Comment by aaronlehmann** [comment]
> Could we add a test case that just calls `compile()` on a bunch of inputs and verifies that it was detected as a correct type. It feels that if there is a breakage in there it could easily go unnoticed because another type could still give a correct match result, but inefficiently.

Added a unit test for `compile`, which verifies the match type, and also checks the assembled regexp in cases where it uses a regexp for matching.

File: pkg/fileutils/fileutils_test.go
Comment (aaronlehmann) [inline pkg/fileutils/fileutils_test.go:708]: We could potentially add match types to cover these cases without regexps, if you think it's worth it.

Comment (aaronlehmann) [reply pkg/fileutils/fileutils_test.go:708]: (They could also cover `*a/b` and `a/b*`)

**Review by tonistiigi — APPROVED** [review]

**Comment by aaronlehmann** [comment]
The CI failure is unrelated - can someone please rerun it and merge?

**Review by thaJeztah — DISMISSED** [review]
LGTM

kicked CI

**Comment by thaJeztah** [comment]
Windows Jenkins machines were fixed, but looks like CI is failing on Windows (looks related);

```
=== RUN   TestCompile/slash
    fileutils_test.go:761: pattern "a*/b": regexp = ^a[^/]*\b$, want ^a[^/]*/b$
    fileutils_test.go:761: pattern "**/**": regexp = ^(.*/)?\*[^/]*$, want ^(.*/)?.*$
    fileutils_test.go:752: pattern "dir/**": matchType = 4, want 2
    fileutils_test.go:752: pattern "**/dir": matchType = 4, want 3
    fileutils_test.go:761: pattern "**/dir2/*": regexp = ^(.*/)?\dir2\*$, want ^(.*/)?dir2/[^/]*$
    fileutils_test.go:761: pattern "**/dir2/**": regexp = ^(.*/)?\dir2\*[^/]*$, want ^(.*/)?dir2/.*$
    fileutils_test.go:761: pattern "**/file*txt": regexp = ^(.*/)?\file[^/]*txt$, want ^(.*/)?file[^/]*txt$
    fileutils_test.go:761: pattern "**/**/*.txt": regexp = ^(.*/)?\*[^/]*\*\.txt$, want ^(.*/)?(.*/)?[^/]*\.txt$
    fileutils_test.go:752: pattern "**/foo/bar": matchType = 4, want 3
    fileutils_test.go:752: pattern "a(b)c/def": matchType = 4, want 1
    fileutils_test.go:749: Failed to compile pattern "dist/proxy.py-2.4.0rc3.dev36+g08acad9-py3-none-any.whl": error parsing regexp: invalid character class range: `\pr`
    --- FAIL: TestCompile/slash (0.00s)
```

**Comment by aaronlehmann** [comment]
The `compile` method was changed to be platform-agnostic, but the test still used `NewPatternMatcher` to create the `PatternMatcher`, which does call `filepath.Clean`. I've changed the test to avoid `NewPatternMatcher` and call `compile` directly on a manually-constructed object. I believe the test will pass on Windows now.

**Comment by thaJeztah** [comment]
Thanks! Those Windows paths are always a PITA.
I'll wait for CI to run, but looks good

**Pull request merged**
